contrib/agent/src/main/java/io/opencensus/contrib/agent/AgentMain.java

```diff
@@ -62,14 +65,48 @@ public static void premain(String agentArgs, Instrumentation instrumentation) th
     instrumentation.appendToBootstrapClassLoaderSearch(
             new JarFile(Resources.getResourceAsTempFile("bootstrap.jar")));
 
+    assertIsLoadedByBootstrapClassloader(ContextManager.class);
+    assertIsLoadedByBootstrapClassloader(ContextStrategy.class);
+
     AgentBuilder agentBuilder = new AgentBuilder.Default()
             .disableClassFormatChanges()
             .with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
             .with(new AgentBuilderListener())
             .ignore(none());
-    // TODO(stschmidt): Add instrumentation for context propagation.
+    agentBuilder = LazyLoaded.addGrpcContextPropagation(agentBuilder);
+    // TODO: Add more instrumentation, potentially introducing a plugin mechanism.
     agentBuilder.installOn(instrumentation);
 
     logger.info("Initialized.");
   }
+
+  private static void assertIsLoadedByBootstrapClassloader(Class<?> clazz) {
+    checkNotNull(clazz, "clazz");
+
+    checkState(clazz.getClassLoader() == null,
+            "%s must be loaded by the bootstrap classloader",
+            clazz);
+  }
+
+  private static class LazyLoaded {
+
+    /**
+     * Adds automatic context propagation.
+     */
+    static AgentBuilder addGrpcContextPropagation(AgentBuilder agentBuilder) {
+      checkNotNull(agentBuilder, "agentBuilder");
+
+      // TODO(stschmidt): Gracefully handle the case of missing io.grpc.Context at runtime.
+
+      ContextManager.setContextHandler(new GrpcContextStrategy());
```

Comment:
Going to add sth. to the javadoc of ContextManager to explain why it is implemented as a singleton.